### PR TITLE
Yahya/features/mock classifier helm chart

### PR DESCRIPTION
**Pull request by yahya130**
This PR addresses creating a reusable CD pipeline which deploys apps to AKS

### CODE REVIEW

File: .github/workflows/kubernetes-cd.yml [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+        description: Specifies the name of the AKS cluster which is required to authenticate to the AKS cluster
+env:
+  helm_chart_directory_path: ./helm-charts
+  temporary_upload_artifact_directory: ./temp/upload/packages
+  temporary_download_artifact_directory: ./temp/download/packages
+  container_registry_domain: ghcr.io
+  organization_domain_name: buerokratt
+  namespace: applications
+
+jobs:
```
Comment (decodingahmed) [inline .github/workflows/kubernetes-cd.yml:34]: There are a few block under here that don't have 2-space indentations. Could you please check that for all lines below here?

File: helm-charts/mock-classifier/templates/ingress.yaml [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+{{- if semverCompare ">=1.19-0" .Capabilities.KubeVersion.GitVersion -}}
+apiVersion: networking.k8s.io/v1
+{{- else if semverCompare ">=1.14-0" .Capabilities.KubeVersion.GitVersion -}}
+apiVersion: networking.k8s.io/v1beta1
+{{- else -}}
+apiVersion: extensions/v1beta1
+{{- end }}
+kind: Ingress
+metadata:
+  name: {{ $fullName }}
```
Comment (decodingahmed) [inline helm-charts/mock-classifier/templates/ingress.yaml:18]: Please add a postfix
```suggestion
  name: {{ $fullName }}-ingress
```

File: helm-charts/mock-classifier/templates/deployment.yaml [hunk 3]
```diff
@@ -0,0 +1,61 @@
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: {{ include "mock-classifier.fullname" . }}
```
Comment (decodingahmed) [inline helm-charts/mock-classifier/templates/deployment.yaml:4]: I know I requested postfixes on the other resources, but please don't add one here.

File: helm-charts/mock-classifier/templates/hpa.yaml [hunk 4]
```diff
@@ -0,0 +1,28 @@
+{{- if .Values.autoscaling.enabled }}
+apiVersion: autoscaling/v2beta1
+kind: HorizontalPodAutoscaler
+metadata:
+  name: {{ include "mock-classifier.fullname" . }}
```
Comment (decodingahmed) [inline helm-charts/mock-classifier/templates/hpa.yaml:5]: Please add `-hpa` postfix so we can easily identify this.
```suggestion
  name: {{ include "mock-classifier.fullname" . }}-hpa
```

File: helm-charts/mock-classifier/templates/service.yaml [hunk 5]
```diff
@@ -0,0 +1,15 @@
+apiVersion: v1
+kind: Service
+metadata:
+  name: {{ include "mock-classifier.fullname" . }}
```
Comment (decodingahmed) [inline helm-charts/mock-classifier/templates/service.yaml:4]:
```suggestion
  name: {{ include "mock-classifier.fullname" . }}-service
```

File: helm-charts/mock-classifier/templates/serviceaccount.yaml [hunk 6]
```diff
@@ -0,0 +1,12 @@
+{{- if .Values.serviceAccount.create -}}
```
Comment (decodingahmed) [inline helm-charts/mock-classifier/templates/serviceaccount.yaml:1]: I didn't think we needed a service account, did we? Please correct me if I've misunderstood.

File: helm-charts/mock-classifier/templates/tests/test-connection.yaml [hunk 7]
```diff
@@ -0,0 +1,15 @@
+apiVersion: v1
```
Comment (decodingahmed) [inline helm-charts/mock-classifier/templates/tests/test-connection.yaml:1]: This is a smart way to test - great idea! 💡

File: helm-charts/mock-classifier/templates/tests/test-connection.yaml [hunk 8]
```diff
@@ -0,0 +1,15 @@
...
+  labels:
+    {{- include "mock-classifier.labels" . | nindent 4 }}
+  annotations:
+    "helm.sh/hook": test
+spec:
+  containers:
+    - name: wget
+      image: busybox
+      command: ['wget']
+      args: ['{{ include "mock-classifier.fullname" . }}:{{ .Values.service.port }}']
```
Comment (decodingahmed) [inline helm-charts/mock-classifier/templates/tests/test-connection.yaml:14]: I don't know how `busybox` works but will this be continuously pinging our service or will it be on a regular but spaced interval?

**Review by decodingahmed — CHANGES_REQUESTED** [review]
This is great work dude. I've added a few suggestions. Let's talk tomorrow about some of the templating in the helm chart as I think we might be able to simplify it quite a bit.

Great work all around 👏

File: .github/workflows/kubernetes-cd.yml [hunk 9]
```diff
@@ -0,0 +1,111 @@
+name: Kubernetes cd
```
[the remark this answers is not in the document]
Comment (keijoraamat) [reply .github/workflows/kubernetes-cd.yml:1]: But is it deploying helm? If helm is package manager for k8s then current file and workflow is quite descriptive. Or perhaps `app-deployment.yml` and name: `Deploy to k8s cluster`

File: helm-charts/mock-classifier/values.yaml [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+    - host: chart-example.local
+      paths:
+        - path: /
+          pathType: ImplementationSpecific
+  tls: []
+  #  - secretName: chart-example-tls
+  #    hosts:
+  #      - chart-example.local
+
+resources: {}
```
Comment (keijoraamat) [inline helm-charts/mock-classifier/values.yaml:59]: I would define limits for resources. If there are some hidden memory leaks and autoscaling maxReplicas is 100 then only thing limiting is the budget..

File: helm-charts/mock-classifier/templates/tests/test-connection.yaml (hunk 8 above)
Comment (keijoraamat) [reply helm-charts/mock-classifier/templates/tests/test-connection.yaml:14]: If I remember correctly then `busybox` is just set of shell utilities used commonly with small distros like Alpine linux. So I'm guessing this test is something like readiness probe.

File: .github/workflows/kubernetes-cd.yml (hunk 9 above)
[the remark this answers is not in the document]
Comment (decodingahmed) [reply .github/workflows/kubernetes-cd.yml:1]: Fair point. I guess I meant, helm chat. But I think what @keijoraamat suggested makes makes better sense 👍

File: helm-charts/mock-classifier/templates/serviceaccount.yaml (hunk 6 above)
Comment (yahya130) [reply helm-charts/mock-classifier/templates/serviceaccount.yaml:1]: we do not, the authentication is done in the pipeline, this was generated automatically during the helm create command I will delete this file

File: helm-charts/mock-classifier/templates/tests/test-connection.yaml (hunk 7 above)
Comment (yahya130) [reply helm-charts/mock-classifier/templates/tests/test-connection.yaml:1]: I'd love to take credit, but this was auto generated 😆 I have added this issue https://github.com/buerokratt/Infrastructure/issues/24 to keep track and build the tests out

File: helm-charts/mock-classifier/templates/tests/test-connection.yaml (hunk 8 above)
Comment (yahya130) [reply helm-charts/mock-classifier/templates/tests/test-connection.yaml:14]: Correct, busy box is a collection of bash utilities, this part was pre-generated, and wget is used as a download command, essentially, I think this is trying to test whether we can download resources from the server, I think this is very similar to the readiness probe essentially, we can just replicate the readiness probe here maybe, there is an issue here to further refine the tests, would be a good discussion https://github.com/buerokratt/Infrastructure/issues/24

File: helm-charts/mock-classifier/values.yaml (hunk 10 above)
Comment (yahya130) [reply helm-charts/mock-classifier/values.yaml:59]: Agreed, there is an item here to further discuss https://github.com/buerokratt/Infrastructure/issues/24, for the sake of the scope of this item I left this out

**Comment by yahya130** [comment]
Thanks @keijoraamat and @decodingahmed for your comments, they were really good and helpful points

**Review by keijoraamat — APPROVED** [review]

**Review by decodingahmed — APPROVED** [review]

**Pull request merged**